### PR TITLE
Add Ruby Discovery Samples Generator

### PR DESCRIPTION
**Pull request by pongad**

### CODE REVIEW

File: src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+class RubyNameProvider {
+
+  private final ImmutableMap<ResourceId, String> NAME_MAP;
+
+  String getName(String apiName, String apiVersion, String resourceName) {
+    ResourceId id = ResourceId.create(apiName, apiVersion, resourceName);
+    return NAME_MAP.get(id);
+  }
+
+  {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java]: convert this to a constructor

File: src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+class RubyNameProvider {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java]: I would prefer this to not have 'Provider' in the name, because we have generally been using that for codegen providers.

File: src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java (hunk 1 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java]: Done.

File: src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+  private final ImmutableMap<ResourceId, String> NAME_MAP;
+
+  String getName(String apiName, String apiVersion, String resourceName) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java]: Why package-private? I don't see a need to hide this from other parts of toolkit.

File: src/main/java/com/google/api/codegen/ruby/RubyProvider.java [hunk 4]
```diff
@@ -57,8 +57,8 @@
   }
 
   @SuppressWarnings("unchecked")
-  public <Element> GeneratedResult generate(
-      Element element, SnippetDescriptor snippetDescriptor, RubyGapicContext context) {
+  public <Element, ContextT> GeneratedResult generate(
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/ruby/RubyProvider.java]: We have been creating a language-level context interface for each language instead of parameterizing the generate function.

File: src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java (hunk 2 above)
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java]: maybe RubyApiaryNameMap

Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java]: Done

File: src/main/java/com/google/api/codegen/ruby/RubyProvider.java (hunk 4 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyProvider.java]: Done.

File: src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java (hunk 3 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyNameProvider.java]: Done.

**Comment by garrettjonesgoogle** [comment]
The generator logic part LGTM. Someone else should review the snippet and generated code.

File: src/test/java/com/google/api/codegen/testdata/discoveries/ruby/ruby_appengine.v1beta5.json.baseline [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+service = Google::Apis::AppengineV1beta5::AppengineService.new
+service.authorization = Google::Auth.get_application_default(['https://www.googleapis.com/auth/cloud-platform'])
+
+# TODO: Change placeholders below to values for parameters to the 'get_app' method:
+# Part of `name`. Name of the application to get. For example: "apps/myapp".
+apps_id = ''
+
+
+# Perform the call
+response = service.get_app(appsId)
```
Comment (jmuk) [inline src/test/java/com/google/api/codegen/testdata/discoveries/ruby/ruby_appengine.v1beta5.json.baseline]: s/appsId/apps_id/ ?

**Comment by jmuk** [comment]
TIL: github gets too slow if too many(?) files are changed in a PR.

Anyways, as far as I see, it looks good to me too, except for camelCase vs snake_case variable names (and Ruby prefers snake_case for local variables).

**Comment by pongad** [comment]
@jmuk Thank you for the review. Unfortunately that camel case thing went past "compile" check, since we can only check initialization. I'll fix this on Monday. We should probably come up with a better strategy for testing dynamic languages 😢

**Comment by pongad** [comment]
@jmuk PTAL

File: src/main/java/com/google/api/codegen/LanguageUtil.java [hunk 6]
```diff
@@ -67,4 +67,8 @@ public static String lowerUnderscoreToLowerCamel(String name) {
   public static String lowerCamelToUpperCamel(String name) {
     return CaseFormat.LOWER_CAMEL.to(CaseFormat.UPPER_CAMEL, name);
   }
+
+  public static String lowerCamelToLowerUnderscore(String name) {
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/LanguageUtil.java]: Could you change the "written  twice" in the class comment above? It seems like now CodegenContext delegates to these static methods, as it should.

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class RubyApiaryNameMap {
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: Class comment

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            StandardCharsets.UTF_8);
+    Map<String, String> nameData = (Map<String, String>) (new Yaml().load(data));
+    ImmutableMap.Builder<ResourceId, String> builder = ImmutableMap.<ResourceId, String>builder();
+    for (Map.Entry<String, String> entry : nameData.entrySet()) {
+      builder.put(parseKey(entry.getKey()), entry.getValue());
+    }
+    return builder.build();
+  }
+
+  private static final Pattern keyPattern = Pattern.compile("^/(.*?):(.*?)[/?](.*)$");
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: comment

File: src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+  private String getNameFromMap(String resourceName) {
+    return apiaryNameMap.getName(getApi().getName(), getApi().getVersion(), resourceName);
+  }
+
+  public String getApiVersion() {
+    return getApi().getVersion().replace('.', '_');
+  }
+
+  private static final ImmutableMap<String, String> SERVICE_RENAME =
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java]: Add a comment: why are these here?

File: src/main/java/com/google/api/codegen/LanguageUtil.java (hunk 6 above)
Comment (jmuk) [inline src/main/java/com/google/api/codegen/LanguageUtil.java]: It's probably out of the scope of this CL, but can we consolidate these two implementations and keep it only in CodegenContext?
LanguageUtil functions are only used by GoGapicContext.java and JavaContextCommon.java, and it's quite easy to rewrite GoGapicContext -- although I'm a bit unsure how JavaContextCommon can be fixed.

File: src/main/java/com/google/api/codegen/ruby/RubyDiscoveryProvider.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Override
+  public GeneratedResult generateFragments(Method method, SnippetDescriptor snippetDescriptor) {
+    return provider.generate(method, snippetDescriptor, context);
+  }
+
+  @Override
+  public void output(String outputPath, Multimap<Method, GeneratedResult> methods)
+      throws IOException {
+    Api api = getService().getApis(0);
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyDiscoveryProvider.java]: Is this the code the same in all languages? In that case we should have this method defined in a common ancestor. If that's complicated because we need to define such an ancestor, we can do that in a follow-up change.

File: src/test/java/com/google/api/codegen/testdata/discoveries/ruby/ruby_appengine.v1beta5.json.baseline [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+# 3. To install the client library and Application Default Credentials library, run:
+#    'gem install google-api-client'
+#    'gem install googleauth'
+require 'googleauth'
+require 'google/apis/appengine_v1beta5'
+
+service = Google::Apis::AppengineV1beta5::AppengineService.new
+service.authorization = Google::Auth.get_application_default(['https://www.googleapis.com/auth/cloud-platform'])
+
+# TODO: Change placeholders below to values for parameters to the 'get_app' method:
```
Comment (vchudnov-g) [inline src/test/java/com/google/api/codegen/testdata/discoveries/ruby/ruby_appengine.v1beta5.json.baseline]: Use the language in the Go comment (corrected) for consistency:

TODO: Change placeholders below to appropriate parameter values for the 'delete' method:

File: src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+            return String.format(
+                "'%s' // 'YYYY-MM-DDThh:mm:ss.fffZ' (UTC)", new DateTime(0L).toStringRfc3339());
+        }
+      }
+      return "''";
+    }
+    return "nil";
+  }
+
+  public String getRequestTypeName(Method method) {
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java]: A comment would be useful--say what the returned request type could be.

File: src/main/java/com/google/api/codegen/LanguageUtil.java (hunk 6 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/LanguageUtil.java]: @vchudnov-g Done.
@jmuk Added issue #122

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java (hunk 7 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: Done.

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java (hunk 8 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: DOne.

File: src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java (hunk 9 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java]: Done

File: src/main/java/com/google/api/codegen/ruby/RubyDiscoveryProvider.java (hunk 10 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyDiscoveryProvider.java]: It is similar in all languages but not the same. It is not too much of an effort to make them all the same though.
Judging from the structure we already have, I think it will take at least two additional classes to move this method in one place. I'm not sure if it's worth the effort.

File: src/test/java/com/google/api/codegen/testdata/discoveries/ruby/ruby_appengine.v1beta5.json.baseline (hunk 11 above)
Comment (pongad) [inline src/test/java/com/google/api/codegen/testdata/discoveries/ruby/ruby_appengine.v1beta5.json.baseline]: Done.

File: src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java (hunk 12 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyDiscoveryContext.java]: Done.

**Comment by pongad** [comment]
@vchudnov-g PTAL

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+import org.yaml.snakeyaml.Yaml;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * RubyApiaryNameMap maps discovery names (found in discovery doc) to user-friendly names (found in
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: "the discovery docs"
"the Ruby client libraries"

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+ * Ruby client libraries). Ruby client library repository has a YAML file "api_names_out.yaml" that
+ * contains this mapping. RubyApiaryNameMap parses the file and provides convenient accessors.
+ */
+public class RubyApiaryNameMap {
+
+  private final ImmutableMap<ResourceId, String> NAME_MAP;
+
+  /**
+   * Constructs RubyApiaryNameMap. It converts IOException into a RuntimeException so that it is
```
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: either
  both "an IOException" + "a RUntimeException"
or
  neither "an" nor "a"

File: src/main/java/com/google/api/codegen/ruby/RubyDiscoveryProvider.java (hunk 10 above)
Comment (vchudnov-g) [inline src/main/java/com/google/api/codegen/ruby/RubyDiscoveryProvider.java]: I suggest creating an issue to track this, and we can do it separately. Repeated code/logic repeatedly bites me.

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java (hunk 13 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: Done.

File: src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java (hunk 14 above)
Comment (pongad) [inline src/main/java/com/google/api/codegen/ruby/RubyApiaryNameMap.java]: Done.

**Comment by vchudnov-g** [comment]
LGTM
